crypto/crypto-derive/src/lib.rs

```diff
@@ -368,6 +375,24 @@ pub fn hasher_dispatch(input: TokenStream) -> TokenStream {
         }
 
         impl libra_crypto::hash::CryptoHasher for #hasher_name {
+            fn seed() -> &'static [u8; 32] {
+                if let Some(seed_value) = #static_seed_name.get() {
+                    seed_value
+                } else {
+                    use libra_crypto::_sha3::Hasher;
+
+                    let name = libra_crypto::_serde_name::trace_name::<#type_name #param>()
+                        .expect("The `CryptoHasher` macro only applies to structs and enums.").as_bytes();
+                    let salt: Vec<u8> = [libra_crypto::hash::LIBRA_HASH_PREFIX, name].concat();
+                    let mut sha3 = libra_crypto::_sha3::Sha3::v256();
+                    sha3.update(&salt);
+                    let mut output = [0u8; 32];
+                    sha3.finalize(&mut output);
```

Comment:
This code is now duplicated in two places: here and in `DefaultHasher::new()`. I feel that we could avoid this and keep LIBRA_HASH_PREFIX `pub(crate)` simply by creating a utility function in `DefaultHasher`.

Reply:
So, it's an excellent suggestion, and I have created a `DefaultHasher::prefixed_hash`, rather than touch `DefaultHasher::new()`.

I do not understand `DefaultHasher::new`, think it is seeding with a double sha-3 of the salt, and is therefore buggy as hell. @kchalkias @valerini @weilei0107 feel free to enlighten me.

Reply:
Good catch! It totally looks like a bug due to the confusing name `HashValue::from_sha3_256`.

Reply:
Thanks for your offline explanation, it seems the `DefaultHasher::new(typename)` is indeed OK, though cryptic.
https://github.com/libra/libra/blob/97c458e54545e01a7f04ee28ea27164368ea20ce/crypto/crypto/src/hash.rs#L484-L488

We start with 
`HashValue::from_sha3_256(&salt[..]) == seed`
then `state.update(seed)`

I think I'll tack on a couple additional commits that rename `from_sha3_256` and and just rephrase the existing `DefaultHasher::new()` in a hopefully more readable fashion.

Reply:
Done! @davidiw I'd love your opinion on whether the last couple of commits in this PR are an improvement.